File: node/test/ledpattern.js
import { KM16 } from "../km16.js";

// Create KM16 instance
let km16 = new KM16();

function sleep(ms) { return new Promise((resolve) => setTimeout(resolve, ms)) }

// Start the watchdog timer with 2 second time out
km16.setWatchDog(2000);

// Current color number for each of LEDs
let indicator = 1;
let underglow = 0;
let keys = 0;

// LED defaults
km16.setLeds(2, mapColorIndexToRGB(indicator));
km16.enableLeds(0, true);
km16.enableLeds(1, true);

let arr = [ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ];

function write_leds()
{
    km16.setLeds(0, arr.map(mapColorIndexToRGB));
}


// Input handler
km16.on("input", (e) => {

    // Log it
    console.log(e);

    // Top left encoder controls underglow LEDs
    if (e.encoder == 1)
    {
        underglow = (underglow + e.delta + 8) % 8;
        km16.setUnderglow(mapColorIndexToRGB(underglow));
    }

    // Top right encoder controls indicator LED
    if (e.encoder == 2)
    {
        indicator = (indicator + e.delta + 8) % 8;
        km16.setIndicator(indicator);
    }
});

function mapColorIndexToRGB(index)
{
    return (index & 0x04 ? 0xFF0000 : 0) | 
            (index & 0x02 ? 0x00FF00 : 0) | 
            (index & 0x01 ? 0x0000FF : 0)
}

function clear()
{
    for (let i=0; i<16; i++)
    {
        arr[i] = 0;
    }
}

function setHBar(y,color)
{
    for (let i=0; i<4; i++)
    {
        arr[y * 4 + i] = color;
    }
}

function setVBar(x, color)
{
    for (let i=0; i<4; i++)
    {
        arr[i * 4 + x] = color;
    }
}


while (true)
{
    let c = 1;
    for (let i=0; i<2; i++)
    {
        for (let col = 0; col < 4; col++)
        {
            clear();
            setVBar(col, c + 1);
            write_leds();   
            await sleep(200);
            c = (c + 1) % 7
        }

        for (let row = 0; row < 4; row++)
        {
            clear();
            setHBar(row, c + 1);
            write_leds();   
            await sleep(200);
            c = (c + 1) % 7
        }
    }

    for (let i=0; i<2; i++)
    {
        for (let i = 0; i < 16; i++)
        {
            clear();
            arr[i] = c + 1;
            write_leds();   
            await sleep(100);
            c = (c + 1) % 7
        }

        for (let i = 14; i > 0; i--)
        {
            clear();
            arr[i] = c + 1;
            write_leds();   
            await sleep(100);
            c = (c + 1) % 7
        }
    }

}

File: python/test/led_pattern.py
import asyncio
import sys
sys.path.insert(0, "..")
from raw_macro_pad import RawMacroPad


def map_color_index_to_rgb(index):
    return ((0xFF0000 if index & 0x04 else 0) |
            (0x00FF00 if index & 0x02 else 0) |
            (0x0000FF if index & 0x01 else 0))


async def main():
    km16 = RawMacroPad()

    # Start the watchdog timer with 2 second timeout
    km16.set_watchdog(2000)

    # LED defaults
    km16.enable_leds(0, True)
    km16.enable_leds(1, True)

    arr = [0] * 16

    def write_leds():
        km16.set_leds(0, [map_color_index_to_rgb(c) for c in arr])

    def clear():
        for i in range(16):
            arr[i] = 0

    def set_h_bar(y, color):
        for i in range(4):
            arr[y * 4 + i] = color

    def set_v_bar(x, color):
        for i in range(4):
            arr[i * 4 + x] = color

    print("Running LED pattern. Press Ctrl+C to exit.")

    try:
        while True:
            c = 1
            for _ in range(2):
                for col in range(4):
                    clear()
                    set_v_bar(col, c + 1)
                    write_leds()
                    await asyncio.sleep(0.2)
                    c = (c + 1) % 7

                for row in range(4):
                    clear()
                    set_h_bar(row, c + 1)
                    write_leds()
                    await asyncio.sleep(0.2)
                    c = (c + 1) % 7

            for _ in range(2):
                for i in range(16):
                    clear()
                    arr[i] = c + 1
                    write_leds()
                    await asyncio.sleep(0.1)
                    c = (c + 1) % 7

                for i in range(14, 0, -1):
                    clear()
                    arr[i] = c + 1
                    write_leds()
                    await asyncio.sleep(0.1)
                    c = (c + 1) % 7

    except (KeyboardInterrupt, asyncio.CancelledError):
        pass
    finally:
        km16.close()


asyncio.run(main())
